haskell-api.md

```diff
@@ -0,0 +1,236 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+Filters should match the dotted-path of the filtered attribute in the response.
+
+For example:
+
+```
+/3/teachers
+[ { id:
+  , ...
+  , school:
+    { id:
+    , ...
+    }
+  }
+]
+```
+
+Here, a School filter would be `school.id`.
+
+```
+/3/students
+[ { id:
+  , ...
+  , schools:
+    [ { id:
+      , ...
+      }
+    ]
+  }
+]
+```
+
+But here, it would be `schools.id`.
+
+If a filter exists that is not for an attribute present in the response, the
+name can be inferred by what it would look like if it were.
```

Comment:
This is super powerful, but could be confusing as well. If a request URL has multiple filters, it might be easy to miss why the response does not have the entities that I would expect. "With great power comes great responsibility."
Also, do we need to build this? Maybe it's already done.

Reply:
Heh, this is not really meant to be any kind of powerful thing... We're just talking about adding some structure around naming choices. This sentence is just to answer the obvious question about a Guide that says to "name parameters like they are in the response", which would be "well, what if it's not in the response?"